File: Exercism Codes/Java/ListOps.java
import java.util.*;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;

class ListOps {

    static <T> List<T> append(List<T> list1, List<T> list2) {
        List<T> newList = new ArrayList<>(list1);
        newList.addAll(list2);
        return newList;
    }

    static <T> List<T> concat(List<List<T>> listOfLists) {
        return listOfLists.stream().flatMap(Collection::stream).toList();
    }

    static <T> List<T> filter(List<T> list, Predicate<T> predicate) {
        return list.stream().filter(predicate).toList();
    }

    static <T> int size(List<T> list) {
        return list.size();
    }

    static <T, U> List<U> map(List<T> list, Function<T, U> transform) {
        return list.stream().map(transform).toList();
    }

    static <T> List<T> reverse(List<T> list) {
        Collections.reverse(list);
        return list;
    }

    static <T, U> U foldLeft(List<T> list, U initial, BiFunction<U, T, U> f) {
        for(T l : list) {
            initial = f.apply(initial, l);
        }
        return initial;
    }

    static <T, U> U foldRight(List<T> list, U initial, BiFunction<T, U, U> f) {
        Collections.reverse(list);
        for(T l : list) {
            initial = f.apply(l, initial);
        }
        return initial;
    }

    private ListOps() {
        // No instances.
    }

}

File: Exercism Codes/Kotlin/ListOps.kt
fun <T> List<T>.customAppend(list: List<T>): List<T> {
    return this.plus(list)
}

fun List<Any>.customConcat(): List<Any> {
    val result = mutableListOf<Any>()
    for (x in this) {
        result.addAll(x as Collection<Any>)
    }
    for (x in result) {
        if (x is List<*>) {
            return result.customConcat()
        }
    }
    return result
}

fun <T> List<T>.customFilter(predicate: (T) -> Boolean): List<T> {
    return this.filter(predicate)
}

val List<Any>.customSize: Int get() = this.size

fun <T, U> List<T>.customMap(transform: (T) -> U): List<U> {
    return this.map(transform)
}

fun <T, U> List<T>.customFoldLeft(initial: U, f: (U, T) -> U): U {
    return this.fold(initial, f)
}

fun <T, U> List<T>.customFoldRight(initial: U, f: (T, U) -> U): U {
    return this.foldRight(initial, f)
}

fun <T> List<T>.customReverse(): List<T> {
    return this.reversed()
}
